### ISSUE TITLE
BIOS support for more RAM

### ISSUE THREAD

**Issue by c0pperdragon** [c0pperdragon/OS816#3]
The OS816 memory architecture allows for fairly easy extension of the RAM to up to 4MB in a modified circuitry. This can easily be done by using a larger static RAM IC. 
But currently there is no direct support from the standard BIOS. There is actually a function call to query the size of the RAM, but this is currently hardcoded to just report 512K. 

@gadyke @HeathenUK
If there is demand for that, I could modify the BIOS code to add some auto-detection to see how much RAM is actually installed and then report his number on the existing bios function call. This would immediately allow user programs to access more dynamic RAM via the malloc/free memory management.

**Comment by gadyke** [comment]
That would be amazing!

I've actually just started to make a list of all the references/constants I will need to update to account for the new memory map and ranges (including, for example, the RAM size function call).

I was also thinking about changing some of the hard coded constants (e.g. for flash location, length) into labels/references so I can define the values at the top of the source rather than mid-file (and sometimes more than once).

Greg

**Comment by gadyke** [comment]
This might be for Issue 1 where we're discussing memory maps, but when you say 4MB SRAM is supported "in a modified circuitry" do you just mean fully wiring up MMU1?

I ask because I think that in every other way you've already got a 4MB address space per-device in MMU1.

**Comment by c0pperdragon** [comment]
Yes, I meant just that, as the current reference schematics does not provide A19-A21

**Comment by gadyke** [comment]
Ah right, understood! Apologies for any confusion, I responded above to your original thought.

**Comment by c0pperdragon** [comment]
I have just uploaded a new revision of the main board sematics and layout + gerbers. It was actually embarasingly easy to add the traces to the PCB layout. I really don't know why I left them out in revision 3.  
This now makes the board much more "hackable" if anyone wants to solder additional parts to it.

**Comment by c0pperdragon** [comment]
The modifications to the bios are done and available in the github releases section.
Autodetection will work for RAM sizes of 512KB, 1MB, 2MB, 4MB.  The detection algorithm assumes that the RAM is always visible in the whole range of 000000 - 3FFFFF. With ICs smaller than 4MB, the same memory locations are visible in the usual repeating pattern.

**Comment by gadyke** [comment]
This is mostly related I guess...I've been thinking about how I intend to lay out the RAM for the board and although 32Mbit SRAM does exist it's expensive (and always SMD, not that that's an issue).

I don't see any design reason why I can't just decode A20 and A21 (either via a '138 with a fixed input or a '139) and spread the RAM over 4 much cheaper 8Mbit chips (using the decoded address as a RAM chip-specific CS#), relying on the pre-existing RAMRD# and RAMWR# signals. For that matter it would still save money (and save a logic IC) just to go for two 16Mbit SRAM chips and use A21 as the CS pin for the two chips.

It seems a pretty standard design choice (it's what Sergey did with the Omega), but please do scream if there's some OS816-specific reason this would be a bad idea!

G

**Comment by c0pperdragon** [comment]
Using 2 ICs to get 4 MByte in total would also have been my choice. 
But you probably need to invert A21 for one of the chips. You can again free up one inverter as discussed earlier.
Or maybe you can get a RAM IC that has two CS inputs (one low-active and one high-active). I remember that once I saw such an arrangement on a 64Kbyte RAM.

**Comment by gadyke** [comment]
Yep that's exactly what I was thinking, most of the 2MByte ICs I've found have both a high and a low CS, including one of the bigger brothers of the 1MByte IC you're already using (AS6C1608)!

And yeah if that doesn't work I can re-purpose an inverter again as you say.

G

**Comment by c0pperdragon** [comment]
I did a quick check on mouser and it seems that there are no parts bigger than 1MByte that run on 5V. So if you want to have 4MB and no voltage translation (which is is real pain, actually), you need 4 ICs. 
On the other hand, if you are content with 2MB this is totally easy as these ICs do have CS# and CS2 inputs.
With 4 parts, address decoding is a bit more complicated and would need at least 2 inverters for A21 and A20.

**Comment by gadyke** [comment]
You're quite right - something I haven't had cause to mention is that I'm planning to run the board on 3.3V (bucked down from a 5V lipo charge/boost board) with a few components powered from the master 5V supply but all **logic** done on 3.3V.

Do shout if there's some obvious reason that wouldn't work - it all seems to be within spec for the W65c816S (I may have to keep to 10-14MHz though).

G

**Comment by c0pperdragon** [comment]
3.3V should also work. I once had a similar board running on 3.3V and I think with 10MHz this is probably OK.
I am mainly using 5V to directly run it from USB.

**Comment by gadyke** [comment]
Yeah I'd be doing the same (and would stick as you say with 4 ICs for RAM decoded via a '138/139), but my grand plan is for this to become a portable device - I'm going to mount a 7" LCD to the PCB driven by an RA8876 (a chip that has roughly the same level of hardware acceleration as a contemporary VDP and plenty of VRAM).

All the I/O on that LCD driver board is 3.3v so although I could translate the voltage back to 5V it's a *lot* of lines to translate (~25) and switching to 3.3V overall opens the door to a bunch of other ICs.

G

**Comment by gadyke** [comment]
I'm also hoping that if I can get my routing and decoupling right and switch to a 4 layer board I might be able to push a 3.3V version to 16MHz without too much effort, but there's a reason I'm making it possible to change out the crystal!
